packages/liveblocks-core/src/lib/objectToQuery.ts

```diff
@@ -0,0 +1,222 @@
+}): string {
+  let filterList: Filter[] = [];
+  const entries = Object.entries(obj);
+
+  /**
+   * Case 1: Simple key value pairs (equal operator)
+   * @example
+   * ```ts
+   * { key: "foo" }
+   * ```
+   */
+  const keyValuePairs = entries.filter(([_, value]) =>
+    isSimpleValue(value)
+  ) as [string, string | number | boolean][];
+
+  filterList = getFiltersFromKeyValuePairs(keyValuePairs);
+
+  /**
+   * Case 2: Key value pairs with operator
+   * @example
+   * ```ts
+   * { key: { startsWith: "fo" } }
+   * ```
+   */
+  const keyValuePairsWithOperator = entries.filter(([_, value]) =>
+    isValueWithOperator(value)
+  ) as [string, Record<"startsWith", string>][];
+
+  filterList = filterList.concat(
+    getFiltersFromKeyValuePairsWithOperator(keyValuePairsWithOperator)
+  );
+
+  /**
+   * Case 3: Indexed key value pairs
+   * @example
+   * ```ts
+   * { key: { nestedKey: "foo" } }
+   * ```
+   * @example
+   * ```ts
+   * { key: { nestedKey: { startsWith: "fo" } } }
+   * ```
+   */
+  const indexedKeys = entries.filter(([_, value]) => {
+    return typeof value === "object" && !("startsWith" in value);
+  });
+
+  indexedKeys.forEach(([key, value]) => {
+    const nestedEntries = Object.entries(value);
+    const nestedKeyValuePairs = nestedEntries
+      .filter(([_, value]) => isSimpleValue(value))
+      .map(([nestedKey, nestedValue]) => [
+        formatFilterKey(key, nestedKey),
+        nestedValue,
+      ]) as [string, string | number | boolean][];
+
+    filterList = filterList.concat(
+      getFiltersFromKeyValuePairs(nestedKeyValuePairs)
+    );
+
+    const nestedKeyValuePairsWithOperator = nestedEntries
+      .filter(([_, value]) => isValueWithOperator(value))
+      .map(([nestedKey, nestedValue]) => [
+        formatFilterKey(key, nestedKey),
+        nestedValue,
+      ]) as [string, { startsWith: string }][];
+
+    filterList = filterList.concat(
+      getFiltersFromKeyValuePairsWithOperator(nestedKeyValuePairsWithOperator)
+    );
+  });
+
+  return filterList
+    .map(({ key, operator, value }) =>
+      formatFilter(key, operator, formatFilterValue(value))
+    )
+    .join(" AND ");
+}
+
+const getFiltersFromKeyValuePairs = (
+  keyValuePairs: [string, string | number | boolean][]
+): Filter[] => {
+  const filters: Filter[] = [];
+  keyValuePairs.forEach(([key, value]) => {
+    filters.push({
+      key,
+      operator: ":",
+      value: value,
+    });
+  });
+
+  return filters;
+};
+
+const getFiltersFromKeyValuePairsWithOperator = (
+  keyValuePairsWithOperator: [string, { startsWith: string }][]
+): Filter[] => {
+  const filters: Filter[] = [];
+  keyValuePairsWithOperator.forEach(([key, value]) => {
+    if ("startsWith" in value && typeof value.startsWith === "string") {
+      filters.push({
+        key,
+        operator: "^",
+        value: value.startsWith,
+      });
+    }
+  });
+
+  return filters;
+};
+
+const isSimpleValue = (value: unknown) => {
```

Comment:
…if you return a type predicate here:

```suggestion
const isSimpleValue = (value: unknown): value is string | number | boolean => {
```

Reply:
In TypeScript 5.5 (yet unreleased), this will get [inferred automatically](https://github.com/microsoft/TypeScript/pull/57465)!

Reply:
Oh nice, about time!